File: benchmarks/network/http_server.lua
#!/usr/bin/env lua
--[[
Simple HTTP server in Lua for benchmarking
Equivalent to the Lisp http_server_simple.lisp
Requires luasocket: luarocks install luasocket
--]]

local socket = require("socket")

local function handle_request(client)
    -- Read request line
    local request_line = client:receive()
    if not request_line then
        return false
    end

    -- Read headers until empty line
    while true do
        local header = client:receive()
        if not header or header == "" then
            break
        end
    end

    -- Send response
    local body = "Hello from Lua!"
    local response = string.format(
        "HTTP/1.1 200 OK\r\n" ..
        "Content-Length: %d\r\n" ..
        "Content-Type: text/plain\r\n" ..
        "\r\n" ..
        "%s",
        #body, body
    )

    client:send(response)
    client:close()
    return true
end

local function main()
    local port = arg[1] or 8080
    local max_requests = 1000

    local server = assert(socket.bind("0.0.0.0", port))
    print("Lua HTTP server listening on port " .. port)

    local count = 0
    while count < max_requests do
        local client = server:accept()
        client:settimeout(10)

        if handle_request(client) then
            count = count + 1
        end
    end

    server:close()
end

main()

File: benchmarks/network/http_server.rb
#!/usr/bin/env ruby
# Simple HTTP server in Ruby for benchmarking
# Equivalent to the Lisp http_server_simple.lisp

require 'socket'

def handle_request(client)
  # Read request line
  request_line = client.gets
  return false unless request_line

  # Read headers until empty line
  loop do
    header = client.gets
    break if header.nil? || header.strip.empty?
  end

  # Send response
  body = "Hello from Ruby!"
  response = "HTTP/1.1 200 OK\r\n" \
             "Content-Length: #{body.bytesize}\r\n" \
             "Content-Type: text/plain\r\n" \
             "\r\n" \
             "#{body}"

  client.write(response)
  client.close
  true
rescue => e
  client.close rescue nil
  false
end

def main
  port = ARGV[0]&.to_i || 8080
  max_requests = 100000

  server = TCPServer.new('0.0.0.0', port)
  puts "Ruby HTTP server listening on port #{port}"

  count = 0
  while count < max_requests
    client = server.accept
    count += 1 if handle_request(client)
  end

  server.close
end

main
